alacritty_terminal/src/term/mod.rs

```diff
@@ -921,6 +1042,107 @@ impl<T> Term<T> {
+        self.add_selection(cols, &self.grid.selection, &mut damage_list);
+        self.add_highlight(cols, &self.last_highlight, &mut damage_list);
+        self.add_highlight(cols, &self.grid.url_highlight, &mut damage_list);
+
+        self.damage = DamageRect::from_point(&self.cursor.point);
+        self.last_selection = self.grid.selection.clone();
+        self.last_highlight = self.grid.url_highlight.clone();
+
+        damage_list
+    }
+
+    /// Add a new position to the current damage, creating and consolidating
+    /// damage rects as needed.
+    fn update_damage(&mut self, line: Line, col: Column) {
+        // Check if our new location is close to our current damage rect, or if
+        // we should instead get our hands on a new one
+        if self.damage.is_close_to_line(line) {
+            self.damage = DamageRect::bounding_rect(&self.damage, &DamageRect {
+                x: min(col.0, self.cursor.point.col.0),
+                y: min(line.0, self.cursor.point.line.0),
+                end_x: max(col.0, self.cursor.point.col.0),
+                end_y: max(line.0, self.cursor.point.line.0),
+            });
+        } else if self.damage_list.len() > 8 {
+            // We have accumulated too many damage rects. Consolidate them
+            // all to clean things up
+            self.damage = self.damage_list.iter().fold(
+                DamageRect::bounding_rect(&self.damage, &DamageRect {
+                    x: min(col.0, self.cursor.point.col.0),
+                    y: min(line.0, self.cursor.point.line.0),
+                    end_x: max(col.0, self.cursor.point.col.0),
+                    end_y: max(line.0, self.cursor.point.line.0),
+                }),
+                |acc, x| DamageRect::bounding_rect(&acc, &x),
+            );
+            self.damage_list.clear();
+        } else {
+            // Try to see if one of our old damage rects are close to our
+            // target. If so, swap that our current damage rect with that
+            // one and update it with our new location.
+            let elem = self.damage_list.iter_mut().find(|e| e.is_close_to_line(line));
+            if let Some(elem) = elem {
+                // An old damage rect was close to our new location. Update
+                // our current damage rect with our previous cursor
+                // position, swap the current rect with the one from the
+                // list, and update the swapped rect with our new position.
+                self.damage = DamageRect::bounding_rect(&self.damage, &DamageRect {
+                    x: self.cursor.point.col.0,
+                    y: self.cursor.point.line.0,
+                    end_x: self.cursor.point.col.0,
+                    end_y: self.cursor.point.line.0,
+                });
+                std::mem::swap(&mut self.damage, elem);
+                self.damage = DamageRect::bounding_rect(&self.damage, &DamageRect {
```

Comment:
This `if` statement is super confusing, since you're using self.damage as a `tmp` buffer. Could you refactor it a bit? You're also setting `self.damage` in the end of both branches, so you can write `self.damage = if let Some(...`.

Reply:
Ugh, that was definitely not the prettiest logic I have ever written. I refactored this function a bit, hope it's a bit more readable. I also swapped the order so that we check if an old damage rect is usable *before* we check if we're out of space in the damage list.